File: 5. Longest Palindromic Substring/LongestPalindromicSubstring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    if (s == null || s.length < 1) return "";
    var start = 0, end = 0;
    for (var i = 0; i < s.length; ++i) {
        var l = Math.max(expandAroundCenter(s, i, i), expandAroundCenter(s, i, i + 1));
        if (l > end - start) {
            start = i - Math.floor((l - 1) / 2);
            end = i + Math.floor(l / 2);
        }
    }
    return s.substring(start, end + 1);
};

var expandAroundCenter = function(s, l, r) {
	while (l >= 0 && r < s.length && s[l] == s[r]) {
        l--;
        r++;
    }
    return r - l - 1;
};

File: 0005. Longest Palindromic Substring/LongestPalindromicSubstring.py
class Solution:
    def expandAroundCenter(self, s, l, r):
    	while (l >= 0 and r < len(s) and s[l] == s[r]):
    		l -= 1
    		r += 1
    	if self.max_len < r - l - 1:
            self.max_len = r - l - 1
            self.start = l + 1
            self.end = r
    
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if (not s or len(s) < 1):
        	return ""
       	self.start = self.end = 0
        self.max_len = 0
        
       	for i in range(len(s)):
       		self.expandAroundCenter(s, i, i)
       		self.expandAroundCenter(s, i, i+1)
       	return s[self.start:self.end]
